src/algebra/module/graded_module.lean

```diff
@@ -0,0 +1,132 @@
+/-
+Copyright (c) 2022 Jujian Zhang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jujian Zhang
+-/
+
+import ring_theory.graded_algebra.basic
+import algebra.direct_sum.decomposition
+
+/-!
+# Graded Module
+
+Given a graded `R`-algebra `A` with grading `𝓐 : ι → submodule R A` and an `A`-module `M` with
+decomposition `ι → addsubmonoid M`,
+we define the typeclass `graded_module 𝓐 𝓜` for internally graded modules.
+
+## Main definitions
+
+* `graded_module 𝓐 𝓜`: an `A`-module `M` is graded by `𝓜` if and only if the decomposition map
+`M → ⨁ i, 𝓜 i` is inverse to the canonical map `⨁ i, 𝓜 i → M`.
+* `graded_module.decompose`: `M` and `⨁ i, 𝓜 i` are isomorphic as `add_comm_monoid`.
+* `graded_module.is_module`: `⨁ i, 𝓜 i` is an `A`-module.
+
+## Tags
+
+graded module
+-/
+
+
+open_locale direct_sum big_operators
+
+variables {ι R A : Type*}
+variables [decidable_eq ι] [add_monoid ι] [comm_semiring R] [semiring A] [algebra R A]
+variables (𝓐 : ι → submodule R A) [graded_algebra 𝓐]
+variables {M : Type*} [add_comm_monoid M] [module A M]
+variables (𝓜 : ι → add_submonoid M)
+
+/--
+Given a graded `R`-algebra `A` graded by `𝓐 : ι → submodule R A` and a decomposition of `A`-module
+`M` into `𝓜 : ι → add_submonoid M`, we say that `M` is graded by `𝓜` if and only if the
+decomposition map `M → ⨁ i, 𝓜 i` is inverse to the canonical map `⨁ i, 𝓜 i → M`.
+-/
+class graded_module extends direct_sum.decomposition 𝓜, set_like.has_graded_smul 𝓐 𝓜
+
+namespace graded_module
+
+
+protected lemma is_internal [graded_module 𝓐 𝓜] : direct_sum.is_internal 𝓜 :=
+direct_sum.decomposition.is_internal _
+
+/--
+If `M` is graded by `𝓜`, then `M` is isomorphic to `⨁ i, 𝓜 i` as `add_comm_monoid`.
+-/
+def decompose [graded_module 𝓐 𝓜] : M ≃+ ⨁ i, 𝓜 i :=
+direct_sum.decompose_add_equiv 𝓜
+
+@[simp] lemma decompose_symm_of [graded_module 𝓐 𝓜] {i : ι} (x : 𝓜 i) :
+  (graded_module.decompose 𝓐 𝓜).symm (direct_sum.of _ i x) = x :=
+direct_sum.coe_add_monoid_hom_of _ _ _
+
+instance self : graded_module 𝓐 (λ i, (𝓐 i).to_add_submonoid) :=
```

Comment:
Maybe `graded_algebra.to_graded_module` would be a good name here to match `semiring.to_module`?

Reply:
done